src/D2L.CodeStyle.Analyzers/ApiUsage/Serialization/ReflectionSerializerAnalyzer.cs

```diff
@@ -0,0 +1,318 @@
+			}
+
+			Diagnostic d = Diagnostic.Create(
+					descriptor,
+					typeDeclaration.Syntax.Identifier.GetLocation()
+				);
+
+			context.ReportDiagnostic( d );
+		}
+
+		private static void ReportMultiplePublicConstructors(
+				SymbolAnalysisContext context,
+				TypeDeclaration typeDeclaration,
+				ImmutableArray<IMethodSymbol> constructors
+			) {
+
+			DiagnosticDescriptor descriptor;
+			switch( typeDeclaration.Kind ) {
+
+				case TypeDeclarationKind.Class:
+					descriptor = ReflectionSerializer_Class_MultiplePublicConstructors;
+					break;
+
+				case TypeDeclarationKind.Record:
+					descriptor = ReflectionSerializer_Record_MultiplePublicConstructors;
+					break;
+
+				default:
+					throw new NotSupportedException( $"Unsupported type declaration kind: { typeDeclaration.Kind }" );
+			}
+
+			for( int i = 1; i < constructors.Length; i++ ) {
+
+				IMethodSymbol constructor = constructors[i];
+				ConstructorDeclarationSyntax declaration = GetFirstDeclaringSyntax<ConstructorDeclarationSyntax>( context, constructor );
+
+				Diagnostic diagnostic = Diagnostic.Create(
+					descriptor: descriptor,
+					location: declaration.Identifier.GetLocation()
+				);
+
+				context.ReportDiagnostic( diagnostic );
+			}
+		}
+
+		private static void ReportConstructorParameterCannotBeDeserialized(
+				SymbolAnalysisContext context,
+				IParameterSymbol parameter
+			) {
+
+			ParameterSyntax declaration = GetFirstDeclaringSyntax<ParameterSyntax>( context, parameter );
+
+			Diagnostic diagnostic = Diagnostic.Create(
+					descriptor: ReflectionSerializer_ConstructorParameterCannotBeDeserialized,
+					location: declaration.Identifier.GetLocation(),
+					messageArgs: new[] { parameter.Name }
+				);
+
+			context.ReportDiagnostic( diagnostic );
+		}
+
+		private static void AnalyzeRecordProperties(
+				SymbolAnalysisContext context,
+				INamedTypeSymbol type
+			) {
+
+			IEnumerable<IPropertySymbol> instanceProperties = type
+				.GetMembers()
+				.Where( m => !m.IsStatic )
+				.OfType<IPropertySymbol>();
+
+			foreach( IPropertySymbol property in instanceProperties ) {
+				AnalyzeRecordProperty( context, property );
+			}
+		}
+
+		private static void AnalyzeRecordProperty(
+				SymbolAnalysisContext context,
+				IPropertySymbol property
+			) {
+
+			if( property.DeclaringSyntaxReferences.IsEmpty ) {
+				return;
+			}
+
+			SyntaxNode declaringSyntax = property
+				.DeclaringSyntaxReferences[0]
+				.GetSyntax( context.CancellationToken );
+
+			if( !( declaringSyntax is PropertyDeclarationSyntax declaration ) ) {
+				return;
+			}
+
+			AccessorListSyntax accessors = declaration.AccessorList;
+			if( accessors == null ) {
+				return;
+			}
+
+			for( int i = 0; i < accessors.Accessors.Count; i++ ) {
+
+				AccessorDeclarationSyntax accessor = accessors.Accessors[i];
+				if( accessor.Kind() == SyntaxKind.InitAccessorDeclaration ) {

```

Comment:
Init-Only setters aren't specific to records they can be used on non-record classes and structs as well.

Reply:
Oh really?  :(

Reply:
Really really

Reply:
Updated